dialects: Extend arith and func

AntonLydike opened the pull request:
This PR adds the following ops:

 - `arith.index_cast`
 - `FuncOp.external` to create functions without a body that are used to inform LLVM of external function definitions.

PapyChacal reviewed (commented):
Could we add tests for IndexCastOp?

AntonLydike commented:
> Could we add tests for IndexCastOp?

Tests for this are coming with the MPI lowering

tobiasgrosser commented:
Is there a specific reason this cannot be tested independently?

AntonLydike commented:
I guess not!

superlopuh commented on tests/dialects/test_func.py:
```diff
@@ -33,7 +33,7 @@ def test_func():
             b := Constant.from_int_and_width(2, i32),
             Addi.get(a, b)
-        ]))   # yapf: disable
+        ]))  # yapf: disable
```
Why disable yapf?

AntonLydike replied:
Because of this:

![image](https://user-images.githubusercontent.com/14147713/220969215-3a39b359-efa9-42dd-a6c8-03b6c11da0eb.png)


It's unreadable.

webmiche replied:
You can sometimes add `# ` at the end of lines to force yapf to format differently. (I would try one before the upper `b` here.)

superlopuh replied:
Ah a comma at the end of `Addi.get(a, b)` will enforce the one line per list element

superlopuh commented on xdsl/dialects/arith.py:
```diff
@@ -630,5 +646,8 @@ def get(operand1: Union[Operation, SSAValue],
+        Maxf,
+
+        # Casts
+        IndexCastOp],
```
feels like a newline here would be good to not have spurious diff on it later

superlopuh commented on xdsl/dialects/arith.py:
```diff
@@ -5,7 +5,7 @@
 from xdsl.dialects.builtin import (ContainerOf, Float16Type, Float64Type, IndexType, IntAttr,
                                    IntegerType, Float32Type, IntegerAttr, FloatAttr,
-                                   Attribute, AnyFloat, AnyIntegerAttr)
+                                   Attribute, AnyFloat, AnyIntegerAttr, Signedness)
```
Would be good to fix this

webmiche commented on xdsl/dialects/arith.py:
```diff
@@ -589,6 +589,22 @@ def get(operand1: Union[Operation, SSAValue],
+class IndexCastOp(Operation):
+    name = "arith.index_cast"
+
+    input: Operand
```
```suggestion
    input: Annotated[Operand, AnyAttr()]
```

superlopuh replied:
I think the recommended way to spell this is `Annotated[Operand, AnyAttr]` without the `()`

webmiche replied:
Not sure whether we changed it already. I am sure we had the discussion lately. I am quite sure though that `()` will work...

AntonLydike replied:
I don't understand why is this the recommended way. Does it provide us with anything? I thought it's cleaner to remove annotations where they don't add anything...

math-fehr replied:
I agree with Anton, we shouldn't add annotations here, because none are needed.
`AnyAttr()` would just add a constraint that's always `True`, so it wouldn't add anything.
I think it's cleaner to have just `Operand`.

webmiche replied:
I agree, but as I argue in the other thread, we should do a conscious decision here, because this is what people will inspect to understand our API and I think we used the Annotated approach in other parts of the repo.

webmiche commented on xdsl/dialects/arith.py:
```diff
@@ -589,6 +589,22 @@ def get(operand1: Union[Operation, SSAValue],
+
+    input: Operand
+
+    result: OpResult
```
```suggestion
    result: Annotated[OpResult, AnyAttr()]
```

webmiche commented on xdsl/dialects/func.py:
```diff
@@ -43,6 +43,19 @@ def from_callable(name: str, input_types: List[Attribute],
         return op
 
+    @staticmethod
+    def external(name: str, input_types: List[Attribute],
```
I am fine with adding this, just for my understanding: This is the same as calling `from_region` with an empty region, right? Or is there another subtle difference I missed? How is this implemented in MLIR, I am wondering?

AntonLydike replied:
Yep, this is basically the same.

math-fehr replied:
Yep, I have no clue why external functions do not have optional region.
Or rather I think I know why. It's probably because before there weren't any notions of optional regions ;)

webmiche reviewed (changes requested):
I am personally unsure about how MLIR handles external regions, so I would prefer an interoperation test case for that.

webmiche commented:
Iis there a reason why you merge both operations in one PR? The PR is still small, so I am fine with it (though it makes the commit history a bit uglier, not that it's not ugly already ;)). Or is there a subtle reason I am missing?

AntonLydike commented:
> Iis there a reason why you merge both operations in one PR? The PR is still small, so I am fine with it (though it makes the commit history a bit uglier, not that it's not ugly already ;)). Or is there a subtle reason I am missing?

I'm preparing a larger PR which contains a lowering from MPI to LLVM + others, and I split off everything that's not necessary to make reviewing easier. I felt that it would be more effort to split these two small changes up (both on the reviewing and on the creating the PR side) than it would be to keep these changes together. I'm fine with either.

AntonLydike commented:
> I am personally unsure about how MLIR handles external regions, so I would prefer an interoperation test case for that.

This is exactly how MLIR handles external functions! But I do agree, that interop tests would be pretty nice here. I have never added interop tests (or interacted with them knowingly), where can I see how they are done?

webmiche commented:
They are mostly lit-based and work by passing the file to mlir-opt, piping it back to xdsl and printing in mlir style and comparing with the initial result. You can find examples for the scf dialect [here]8https://github.com/xdslproject/xdsl/blob/main/tests/filecheck/mlir-conversion/with-mlir/dialects/scf/for_mlir.mlir).

math-fehr reviewed (approved):
Nice! That looks good to me!
Just try the `#` or the `,` technique on the format problem, and then I"m good to merge it!

AntonLydike commented:
I just want to add, that while the trailing comma is a good solution, I feel that having each line end in a `# ` is much more problematic than a `# yapf: disable` because it achieves the same thing while being much more invasive and less descriptive.

AntonLydike commented:
@webmiche I removed the `yapf: disable` from the tests (which funnily enough wasn't even added in this PR).

I have not changed the `Annotated[..., AnyAttr()]` thing as @math-fehr said it's okay. I removed it initially on his recommendation, and I would like that discussion to complete before I change it back. I opened a discussion on zulip where we can discuss this. Once that discussions resolves I'll open a separate PR changing it to whatever format we agree on.
